src/plugins/workspace/server/workspace_client.ts

```diff
@@ -260,11 +280,37 @@ export class WorkspaceClient implements IWorkspaceClientImpl {
             );
           }
         }
-        if (promises.length > 0) {
-          await Promise.all(promises);
+      }
+
+      if (newDataConnections) {
+        const originalSelectedDataConnectionIds = originalSelectedDataSourcesAndConnections
+          .filter((item) => item.type === DATA_CONNECTION_SAVED_OBJECT_TYPE)
+          .map((ds) => ds.id);
+        const dataConnectionsToBeRemoved = originalSelectedDataConnectionIds.filter(
+          (ds) => !newDataConnections.find((item) => item === ds)
+        );
+        const dataConnectionsToBeAdded = newDataConnections.filter(
+          (ds) => !originalSelectedDataConnectionIds.find((item) => item === ds)
+        );
+        if (dataConnectionsToBeRemoved.length > 0) {
+          for (const dataConnectionId of dataConnectionsToBeRemoved) {
+            promises.push(
+              client.deleteFromWorkspaces(DATA_CONNECTION_SAVED_OBJECT_TYPE, dataConnectionId, [id])
+            );
+          }
+        }
+        if (dataConnectionsToBeAdded.length > 0) {
+          for (const dataConnectionId of dataConnectionsToBeAdded) {
+            promises.push(
+              client.addToWorkspaces(DATA_CONNECTION_SAVED_OBJECT_TYPE, dataConnectionId, [id])
+            );
+          }
         }
       }
```

Comment:
Nit: seems like the logic to handle the data connection are the same as data source, would it be possible to make it into a function like `addOperationsToPromise(type: string)`?

So the code here will be like
```
const promises = [];
const addOperationsToPromise = (type: string) => { xxx }

// Add data sources operations into promise
addOperationsToPromise(DATA_SOURCE_SAVED_OBJECT_TYPE);

// Add data connections operations into promise
addOperationsToPromise(DATA_CONNECTION_SAVED_OBJECT_TYPE);
```

Reply:
Sounds good, will optimize this part in https://github.com/opensearch-project/OpenSearch-Dashboards/pull/8013.